tests/test_api.rs

```diff
@@ -1036,58 +1035,39 @@ fn promise_rejected() {
   drop(locker);
 }
 
-extern "C" fn fn_callback(info: &v8::FunctionCallbackInfo) {
-  assert_eq!(info.length(), 0);
-  {
-    let rv = &mut info.get_return_value();
-    #[allow(mutable_transmutes)]
-    #[allow(clippy::transmute_ptr_to_ptr)]
-    let info: &mut v8::FunctionCallbackInfo =
-      unsafe { std::mem::transmute(info) };
-    {
-      let mut hs = v8::HandleScope::new(info);
-      let scope = hs.enter();
-      let s = v8::String::new(scope, "Hello callback!").unwrap();
-      let value: v8::Local<v8::Value> = s.into();
-      let rv_value = rv.get(scope);
-      assert!(rv_value.is_undefined());
-      rv.set(value);
-    }
-  }
+fn fn_callback(
+  scope: v8::FunctionScope,
+  args: v8::FunctionArguments,
+  mut rv: v8::ReturnValue,
+) {
+  assert_eq!(args.length(), 0);
+  let s = v8::String::new(scope, "Hello callback!").unwrap();
+  let rv_value = rv.get(scope);
+  assert!(rv_value.is_undefined());
+  rv.set(s.into());
 }
 
-extern "C" fn fn_callback2(info: &v8::FunctionCallbackInfo) {
-  #[allow(mutable_transmutes)]
-  #[allow(clippy::transmute_ptr_to_ptr)]
-  let info: &mut v8::FunctionCallbackInfo =
-    unsafe { std::mem::transmute(info) };
-  assert_eq!(info.length(), 2);
-  let isolate = info.get_isolate();
-  let mut locker = v8::Locker::new(&isolate);
-  let mut context = isolate.get_current_context();
-  let arg1 = info.get_argument(0);
-  let arg2 = info.get_argument(1);
-  let rv = &mut info.get_return_value();
-  {
-    let mut hs = v8::HandleScope::new(&mut locker);
-    let scope = hs.enter();
-    context.enter();
-
-    let arg1_val = v8::String::new(scope, "arg1").unwrap();
-    assert!(arg1.is_string());
-    assert!(arg1.strict_equals(arg1_val.into()));
-
-    let arg2_val = v8::Integer::new(scope, 2);
-    assert!(arg2.is_number());
-    assert!(arg2.strict_equals(arg2_val.into()));
-
-    let s = v8::String::new(scope, "Hello callback!").unwrap();
-    let value: v8::Local<v8::Value> = s.into();
-    let rv_value = rv.get(scope);
-    assert!(rv_value.is_undefined());
-    rv.set(value);
-    context.exit();
-  }
+fn fn_callback2(
+  scope: v8::FunctionScope,
```

Comment:
@bartlomieju, @ry 

You guys ok if I break up `FunctionCallbackInfo` into three pieces as demonstrated here. That makes it a lot easier to play nice with the rust borrow checker.

Reply:
Does this mean overhead for functions that wouldn’t otherwise use the return value ?

Reply:
Does this mean overhead for functions that wouldn’t otherwise use the return value ? EG Deno.core.print

Reply:
Also worth noting that this brings the API further away from the C++ API

Reply:
> Does this mean overhead for functions that wouldn’t otherwise use the return value ? EG Deno.core.print

No this is a zero cost abstraction.

Reply:
> Also worth noting that this brings the API further away from the C++ API

I am well aware. I try to avoid it as much as possible.

Reply:
GetReturnValue isn’t called on every call?

Reply:
> GetReturnValue isn’t called on every call?

It does. However GetReturnValue() does basically nothing, it just returns a pointer from the FunctionCallbackInfo struct. The status quo where we have to create an EscapableHandleScope in all these function callbacks probably wastes more cycles (although, in the grand scheme of things, just as insignificant).

The ReturnValue argument can also be made optional if so desired.

Reply:
Ok so it’s not really zero cost - we’re forcing the user to make more V8 API calls than necessary. But maybe the small overhead is worth it.

The EscapableHandleScope was also overhead induced by the type system. So it’s good you’re reducing the overhead.